### ISSUE TITLE
Compilation cache can lead to segfault

### ISSUE THREAD

**Issue by elmattic** [HigherOrderCO/Kind#228]
### Describe the bug
Running code through the compilation cache can lead to segfault. Running the code directly without type checking it first works.

### To Reproduce
1. Clone my [fork](https://github.com/elmattic/Kind) and checkout my `elmattic/wasm` branch and `cd` to `base`.
2. Type check code with: `kind Main.kind`
3. Run with: `kind Main --run`

Output:

```
base$ kind Main --run
zsh: segmentation fault  kind Main --run
```

### Expected behavior
This should give same result than without using cache:

```
base$ rm -Rf .cache
base$ time kind Main --run
id: 1#8
section id: 1#8, size:52#32
id: 2#8
section id: 2#8, size:66#32
id: 3#8
section id: 3#8, size:16#32
id: 4#8
section id: 4#8, size:5#32
id: 5#8
section id: 5#8, size:3#32
id: 6#8
section id: 6#8, size:21#32
id: 7#8
section id: 7#8, size:55#32
id: 9#8
id: 9#8
section id: 9#8, size:10#32
id: 10#8
Invalid instruction 28
Expected byte 0b but read 28
Invalid instruction 28
Expected byte 0b but read 28
Expected byte 0b but read 28
Expected byte 0b but read 28
oi
kind Main --run  9.90s user 0.19s system 113% cpu 8.875 total
```

### Environment:
 - OS: `macOS 10.15.7`
 - Kind version: `1.0.72`

### Additional context

Note that the errors during reading Wasm hex string are normal (it's a wip and some instructions have been commented).

Also the string sample `Wasm.test.sample.1` works perfectly even through the cache:

```
base$ rm -Rf .cache  
base$ kind Main.kind 
Main: IO(Unit)
Wasm.parse: (hexstr:String) Pair(Buffer8,U32)
Wasm.parse.go: (bc:String) (i:U32) (buff:Buffer8) Buffer8
Wasm.test.sample.1: String
Wasm.test.sample.2: String

All terms check.
base$ kind Main --run
...
(module
  (table $table0 1 1 funcref)
  (memory $memory0 2)
  (global $global0 (mut i32) (i32.const 66560))
  (global $global1 i32 (i32.const 66560))
  (global $global2 i32 (i32.const 1024))
  (export "memory" (memory $memory0))
  (export "__heap_base" (global $global1))
  (export "__data_end" (global $global2))
  (export "main" (func $func1))
  (func $func0
  )
  (func $func1 (param $var0 i32) (param $var1 i32) (result i32)
    (local $var2 i32) (local $var3 i32)
    block $label0
      local.get $var0
      i32.eqz
      br_if $label0
      i32.const 1
      local.set $var2
      loop $label1
        local.get $var2
        local.get $var0
        i32.mul
        local.set $var2
        local.get $var0
        i32.const 127
        i32.add
        local.tee $var3
        local.set $var0
        local.get $var3
        br_if $label0
      end $label1
      local.get $var2
      return
    end $label0
    i32.const 1
  )
)
oi
```

#### Core dump

https://github.com/uwu-tech/Kind/files/6729171/dump.log

**Comment by rigille** [comment]
Sorry for taking so long to answer! That's a stack overflow. The quickest fix is to use the scheme release.
You're reading a fairly large string, and that's ok without the cache because Kind's string parser is tail-recursive. The cache however is stored as FormCore, and FormCore's string parser isn't tail-recursive. I'll implement a more efficient FormCore parser.

**Comment by rigille** [comment]
Also that's a nice wasm parser the one you got in your fork

**Comment by rigille** [comment]
closed by 3e4a71b
